### DISCUSSION TITLE
Does SQLLab / Query Editor support Row Level Security?

### DISCUSSION THREAD

**Discussion by srinisubramanian** [apache/superset#18632] [Q&A / Help] [answered]
I created a datasource for a table and limited the data with RLS. This works fine when I use the dataset to design the dashboard.

When I work with SQLLab (The query editor) and open the same table as in the dataset, the entire data set is now displayed. The RLS does not seem to limit the data for SQLLab. Is that how its expected to work or is this an issue to be resolved?

**Comment by srinisubramanian** [comment]
@srinify as requested moved it here .. but no answers yet. Any help is appreciated

**Comment by villebro** [comment, marked as answer]
@srinisubramanian RLS does not affect SQL Lab currently, and in my experience RLS will usually be used mostly for users that don't have SQL Lab access and are primarily dashboard viewers/dataset explorers as opposed to SQL Lab analysts or virtual table authors.

**Reply by srinisubramanian** [reply]
OK .. the need was to setup Superset in a multi-tenant mode.  While consumers of dashboards will now be able to see only their data via RLS, those who design dashboards will be able to see all data via SQLLab which I wanted to avoid.  Any idea how that can be achieved @villebro ?  Is it even possible?

**Comment by villebro** [comment]
@srinisubramanian I think it will be very difficult to apply RLS to queries in SQL Lab, so I don't really see any easy way to achieve it. But I'm open to considering it if it's possible. Do you have some ideas how it could be achieved in your context?

**Reply by n614cd** [reply]
I can see two approaches:
1. Any table which requires RLS would have to have a common column name or condition. RLS would require a second field which lists the columns that must exist on the table to include the "clause" section.
2. Have the ability to create a separate clause statement per table instead of a single global statement.

With either solution, the RLS clause would be inlined in the from statement and aliased the table name. This is supported by most databases. Otherwise, SQLLab would require fairly complex logic to handle table alias names, or a table alias naming convention for the clause for which the individual writing the clause would need to be aware.

Tim

**Comment by srinisubramanian** [comment]
No. I also could not think of any easy ideas :)  the only way for now is to
give the customer a separate db without multi tenancy.

**Comment by srinisubramanian** [comment]
Would it work if views are created and the db user is restricted to only
the views and not the database?
